Create rule S7044(JS): Server-side requests should not be vulnerable to traversing attacks APPSEC-2042 (#4175)
SonarSource/rspec: rules/S7044/javascript/how-to-fix-it/node.adoc

== How to fix it in Node.js

=== Code examples

include::../../common/fix/code-rationale.adoc[]

==== Noncompliant code example

[source,javascript,diff-id=1,diff-type=noncompliant]
----
const axios = require('axios');
const express = require('express');

const app = express();

app.get('/example', async (req, res) => {
    const target = req.query.target;

    try {
        await axios.get($`https://example.com/user/{target}`); // Noncompliant
        res.send("OK");
    } catch (err) {
        console.error(err);
        res.send("ERROR");
    }
})
----

==== Compliant solution

[source,javascript,diff-id=1,diff-type=compliant]
----
const axios = require('axios');
const express = require('express');

> 

Reviewer: ```suggestion

const app = express();

```

app.get('/example', async (req, res) => {
    const target = (EncodeURLComponent(req.query.target));

    try {
        await axios.get($`https://example.com/user/{target}`);
        res.send("OK");
    } catch (err) {
        console.error(err);
        res.send("ERROR");
    }
})
----

=== How does this work?

> include::../../common/fix/encoding.adoc[]

Reviewer: ```suggestion
include::../../common/fix/encoding.adoc[]

Note that https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI[`encodeURI()`] does not encode enough to be considered a protection against this vulnerability.
```

== How to fix it in Node.js

=== Code examples

include::../../common/fix/code-rationale.adoc[]

==== Noncompliant code example

[source,javascript,diff-id=1,diff-type=noncompliant]
----
const axios = require('axios');
const express = require('express');

const app = express();

app.get('/example', async (req, res) => {
    const id = req.query.id;

    try {

>         await axios.get($`https://example.com/user/{id}`); // Noncompliant

Reviewer: ```suggestion
        await axios.get(`https://example.com/user/{id}`); // Noncompliant
```

>     const target = req.query.target;
> 
>     try {
>         await axios.get($`https://example.com/user/{target}`); // Noncompliant

Reviewer: ```suggestion
    const id = req.query.id;

    try {
        await axios.get($`https://example.com/user/{id}`); // Noncompliant
```

        res.send("OK");
    } catch (err) {
        console.error(err);
        res.send("ERROR");
    }
})
----

==== Compliant solution

[source,javascript,diff-id=1,diff-type=compliant]
----
const axios = require('axios');
const express = require('express');

const app = express();

app.get('/example', async (req, res) => {
    const id = EncodeURIComponent(req.query.id);

    try {

>         await axios.get($`https://example.com/user/?id={id}`);

Reviewer: ```suggestion
        await axios.get(`https://example.com/user/?id={id}`);
```

>         await axios.get($`https://example.com/user/{target}`);

Reviewer: ```suggestion
        await axios.get($`https://example.com/user/?target={target}`);
```

>     const target = EncodeURIComponent(req.query.target);
> 
>     try {
>         await axios.get($`https://example.com/user/?target={target}`);
>         res.send("OK");

Reviewer: ```suggestion
    const id = EncodeURIComponent(req.query.id);

    try {
        await axios.get($`https://example.com/user/?id={id}`);
        res.send("OK");
```

    } catch (err) {
        console.error(err);
        res.send("ERROR");
    }
})
----

=== How does this work?

include::../../common/fix/encoding.adoc[]

> Note that https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI[`encodeURI()`] does not encode enough to be considered a protection against this vulnerability.

Reviewer 2: ```suggestion
Note that https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI[`encodeURI()`] does not encode forward slashes and can therefore not prevent this vulnerabilty.
```
This is just a suggestion. I was thinking that it might be nice to explicitly mention which character is not encoded.
Up to you :)
